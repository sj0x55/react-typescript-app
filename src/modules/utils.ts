export const compose = <T extends unknown[], R>(fn1: (...args: T) => R, ...fns: Array<(a: R) => R>) => {
  const piped = fns.reduce(
    (prevFn, nextFn) => (value: R) => nextFn(prevFn(value)),
    (val) => val,
  );
  return (...args: T) => piped(fn1(...args));
};
