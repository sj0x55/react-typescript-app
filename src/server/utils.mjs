import terminus from '@godaddy/terminus';
import http from 'http';
import kill from 'kill-port';
import axios from 'axios';
import cheerio from 'cheerio';
import userAgents from './user-agents.mjs';

export async function killPort(port) {
  if (process.env.NODE_ENV === 'development') {
    try {
      await kill(port); // After re-run the server during the watch mode the port may be busy.
    } catch (err) {
      console.error(err);
    }
  }
}

export function getCorsConfig(ipAddress) {
  return {
    origin: (origin, callback) => {
      if (new RegExp(`(http|http)://(localhost|127.0.0.1|${ipAddress})`).test(origin)) {
        callback(null, [origin]);
      }
    },
    methods: ['GET', 'POST', 'PUT', 'PATCH', 'DELETE', 'OPTIONS'],
  };
}

export async function listenServer(app, port, callback) {
  const server = http.createServer(app);
  const onSignal = async () => {
    console.log('Server is starting cleanup');
  };
  const onShutdown = async () => {
    console.log('Cleanup finished, server is shutting down');
  };

  terminus.createTerminus(server, { signal: 'SIGINT', onSignal, onShutdown, logger: console.error });
  server.listen(port, '0.0.0.0', callback);
}

export async function fetchContent(url) {
  try {
    const userAgent = userAgents[Math.floor(Math.random() * userAgents.length)];
    const result = await axios.get(url, {
      params: {
        t: new Date().getTime(),
      },
      headers: {
        Pragma: 'no-cache',
        Expires: '0',
        'Cache-Control': 'no-cache',
        'User-Agent': userAgent,
      },
    });

    return await cheerio.load(result.data);
  } catch (err) {
    console.error(err);
    return await cheerio.load('');
  }
}

export async function fetchAllDataFromWishList(baseUrl, pathUrl) {
  const parsePrice = (price) => (price && price.trim().replace(/[^0-9.]+/, '')) || 0;
  const prepareProduct = (product) => (condition, price) => ({
    ...product,
    condition,
    price: parseFloat(price),
  });
  const data = [];
  const content = await fetchContent(`${baseUrl}${pathUrl}`);

  content('[id^="itemInfo_"]').each((i, item) => {
    const dataInfoElement = content(item).find('[data-item-prime-info]');
    const { id, asin } = JSON.parse(dataInfoElement.attr('data-item-prime-info') || '{}');

    if (asin) {
      const title = content(item).find(`#itemName_${id}`).text().trim();
      const href = content(item).find(`#itemName_${id}`).attr('href').trim().split('?')[0];
      const priceForNew = parsePrice(content(item).find(`#itemPrice_${id} > span:first-child`).text());
      const priceForUsed = parsePrice(content(item).find('.itemUsedAndNew > .itemUsedAndNewPrice').text());
      const capacity = content(item)
        .find('span:not(:has(*))')
        .map((j, span) => {
          const spanText = content(span).text().replace(/\s+/g, ' ').trim();
          const matchedCapacity = spanText.match(/^Capacity\s?:\s?(\d+)\s?TB/i);

          return (matchedCapacity && parseInt(matchedCapacity[1])) || null;
        })
        .toArray()
        .find(Number);

      if (capacity) {
        const cProduct = prepareProduct({
          url: `${baseUrl}${href}`,
          title,
          capacity,
        });

        if (priceForNew) {
          data.push(cProduct('new', priceForNew));
        }

        if (priceForUsed) {
          data.push(cProduct('used', priceForUsed));
        }
      }
    }
  });

  return data;
}

export const arrayToObject = (propKeyNames, items) => {
  const obj = {};
  const genKey = (item) =>
    [propKeyNames]
      .flat()
      .map((key) => (item[key] || '').replace(/[^a-z0-9]+/gi, '-'))
      .filter((val) => val)
      .join('-');

  items.forEach((item, index) => {
    obj[genKey(item)] = { ...item, index };
  });

  return obj;
};
