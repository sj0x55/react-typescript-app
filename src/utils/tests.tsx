import { FC, ReactElement, ReactNode } from 'react';
import { Provider } from 'react-redux';
import { ThemeProvider } from 'styled-components';
import { render } from '@testing-library/react';
import testRenderer, { TestRendererOptions } from 'react-test-renderer';
import { configureStore, EnhancedStore } from '@reduxjs/toolkit';
import { mainTheme } from 'styles/theme';
import appSliceReducer from 'app/slice';

type RenderOptions = {
  store?: EnhancedStore<{ app: AppState }>;
  [key: string]: unknown;
};

const getWrapper = (
  store: EnhancedStore<{ app: AppState }> = configureStore({
    reducer: {
      app: appSliceReducer,
    },
  }),
): FC => {
  const wrapper: FC = ({ children }) => (
    <Provider store={store}>
      <ThemeProvider theme={mainTheme}>{children}</ThemeProvider>
    </Provider>
  );

  return wrapper;
};

const customRender = (ui: ReactElement, opts?: RenderOptions) => {
  const renderResult = render(ui, { wrapper: getWrapper((opts || {}).store), ...opts });
  const firstChild = renderResult.container.firstChild;

  return { ...renderResult, firstChild };
};

const customCreate = (ui: ReactElement, opts?: TestRendererOptions) => {
  const Wrapper = getWrapper();

  return testRenderer.create(<Wrapper>{ui}</Wrapper>, opts);
};

const mockComponent = function () {
  return ({ children, ...props }: { children: ReactNode }) => {
    return <span {...props}>{children}</span>;
  };
};

export * from '@testing-library/react';
export { customCreate as create };
export { customRender as render };
export { mockComponent };
