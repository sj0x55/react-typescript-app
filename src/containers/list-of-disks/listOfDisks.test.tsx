import { render } from 'utils/tests';
import { ListOfDisks } from '.';

describe('ListOfDisks component', () => {
  it('should render list with few items', () => {
    const data: DiskItem[] = [
      {
        title: 'WD 5 TB My Passport',
        image: 'https://image1',
        href: 'https://url1',
        condition: 'new',
        features: {
          capacity: 5,
        },
        price: 120.59,
        isPriceChanged: false,
        isPriceDropped: false,
        timestampChanged: 1632167550962,
        prevPrice: null,
        pricePerTB: 24.11,
        prevPricePerTB: null,
      },
      {
        title: 'WD 5 TB My Passport',
        image: 'https://image1',
        href: 'https://url1',
        condition: 'used',
        features: {
          capacity: 5,
        },
        price: 110.59,
        isPriceChanged: false,
        isPriceDropped: false,
        timestampChanged: 1632167550962,
        prevPrice: null,
        pricePerTB: 22.11,
        prevPricePerTB: null,
      },
      {
        title: 'Seagate 8 TB',
        image: 'https://image2',
        href: 'https://url2',
        condition: 'used',
        features: {
          capacity: 8,
        },
        price: 180.9911,
        isPriceChanged: true,
        isPriceDropped: true,
        timestampChanged: 1632167550962,
        prevPrice: 219.9911,
        pricePerTB: 22.6211,
        prevPricePerTB: 27.4911,
      },
      {
        title: 'Seagate 8 TB (no image, no date)',
        image: null,
        href: 'https://url2',
        condition: 'used',
        features: {
          capacity: 8,
        },
        price: 180.99,
        isPriceChanged: true,
        isPriceDropped: true,
        timestampChanged: 0,
        prevPrice: 219.99,
        pricePerTB: 22.62,
        prevPricePerTB: 27.49,
      },
    ];

    const { firstChild } = render(<ListOfDisks data={data} />);

    expect(firstChild).toMatchSnapshot();
  });
});

// describe('dynamicList operations', () => {
//   it('should return default value for empty value', () => {
//     expect(prepareValue('', { prefix: '$ ', suffix: ' #' })).toBe('-');
//     expect(prepareValue(' ', { prefix: '$ ', suffix: ' #' })).toBe('-');
//   });

//   it('should return a value when used a number', () => {
//     expect(prepareValue(123)).toBe('123');
//     expect(prepareValue(123.99)).toBe('123.99');
//   });

//   it('should return a value when used a string', () => {
//     expect(prepareValue('test-value')).toBe('test-value');
//     expect(prepareValue(' test-value ')).toBe('test-value');
//     expect(prepareValue('test-value  with-space')).toBe('test-value  with-space');
//   });

//   it('should return a value with prefix and suffix when used a string', () => {
//     expect(prepareValue('test-value', { prefix: '$ ', suffix: ' #' })).toBe('$ test-value #');
//   });
// });
