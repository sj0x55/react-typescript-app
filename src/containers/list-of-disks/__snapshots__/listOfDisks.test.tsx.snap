// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListOfDisks component should render list with few items 1`] = `
.c0 {
  display: grid;
}

.c3 {
  text-align: center;
}

.c6 {
  text-align: right;
}

.c1 {
  border: 1px solid #66748f;
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 10px 10px;
  border-top: 1px solid #C0C7D6;
}

.c4:nth-child(16n + 1) {
  background: #e5e8ee;
}

.c4:nth-child(16n + 2) {
  background: #e5e8ee;
}

.c4:nth-child(16n + 3) {
  background: #e5e8ee;
}

.c4:nth-child(16n + 4) {
  background: #e5e8ee;
}

.c4:nth-child(16n + 5) {
  background: #e5e8ee;
}

.c4:nth-child(16n + 6) {
  background: #e5e8ee;
}

.c4:nth-child(16n + 7) {
  background: #e5e8ee;
}

.c4:nth-child(16n + 8) {
  background: #e5e8ee;
}

.c4:nth-child(-n + 8) {
  background: #66748f;
  color: #FFFFFF;
  border-top: none;
  border-bottom: none;
}

.c5 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 10px 10px;
  border-top: 1px solid #C0C7D6;
}

.c5:nth-child(16n + 1) {
  background: #e5e8ee;
}

.c5:nth-child(16n + 2) {
  background: #e5e8ee;
}

.c5:nth-child(16n + 3) {
  background: #e5e8ee;
}

.c5:nth-child(16n + 4) {
  background: #e5e8ee;
}

.c5:nth-child(16n + 5) {
  background: #e5e8ee;
}

.c5:nth-child(16n + 6) {
  background: #e5e8ee;
}

.c5:nth-child(16n + 7) {
  background: #e5e8ee;
}

.c5:nth-child(16n + 8) {
  background: #e5e8ee;
}

.c5:nth-child(-n + 8) {
  background: #66748f;
  color: #FFFFFF;
  border-top: none;
  border-bottom: none;
}

.c7 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: right;
  -webkit-justify-content: right;
  -ms-flex-pack: right;
  justify-content: right;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 10px 10px;
  border-top: 1px solid #C0C7D6;
}

.c7:nth-child(16n + 1) {
  background: #e5e8ee;
}

.c7:nth-child(16n + 2) {
  background: #e5e8ee;
}

.c7:nth-child(16n + 3) {
  background: #e5e8ee;
}

.c7:nth-child(16n + 4) {
  background: #e5e8ee;
}

.c7:nth-child(16n + 5) {
  background: #e5e8ee;
}

.c7:nth-child(16n + 6) {
  background: #e5e8ee;
}

.c7:nth-child(16n + 7) {
  background: #e5e8ee;
}

.c7:nth-child(16n + 8) {
  background: #e5e8ee;
}

.c7:nth-child(-n + 8) {
  background: #66748f;
  color: #FFFFFF;
  border-top: none;
  border-bottom: none;
}

.c10 {
  font-size: 12px;
}

.c9 {
  cursor: pointer;
  -webkit-text-decoration: none;
  text-decoration: none;
  color: #66748f;
}

.c9:hover {
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

.c8 {
  border: 2px solid #C0C7D6;
}

.c2 {
  grid-template-columns: 40px 120px 200px 200px 150px 150px 1fr 120px;
}

<div
  class="c0 c1 c2"
  data-test="dynamicData"
  sizes="40px,120px,200px,200px,150px,150px,1fr,120px"
>
  <div
    class="c3 c4"
  >
    -
  </div>
  <div
    class="c3 c4"
  >
    Image
  </div>
  <div
    class="c3 c4"
  >
    Price per TB
  </div>
  <div
    class="c3 c4"
  >
    Price
  </div>
  <div
    class="c3 c4"
  >
    Capacity
  </div>
  <div
    class="c3 c4"
  >
    Condition
  </div>
  <div
    class="c5"
  >
    Name
  </div>
  <div
    class="c6 c7"
  >
    Last update
  </div>
  <div
    class="c3 c4"
    data-condition="used"
    data-test="cell0-1"
  >
    <span
      class=" "
    >
      -
    </span>
  </div>
  <div
    class="c3 c4"
    data-condition="used"
    data-test="cell0-2"
  >
    <img
      class="c8"
      height="60"
      src="https://image1"
      width="60"
    />
  </div>
  <div
    class="c3 c4"
    data-condition="used"
    data-test="cell0-3"
  >
    <span
      class=" "
    >
      £ 22.12
    </span>
  </div>
  <div
    class="c3 c4"
    data-condition="used"
    data-test="cell0-4"
  >
    <span
      class=" "
    >
      £ 110.59
    </span>
  </div>
  <div
    class="c3 c4"
    data-condition="used"
    data-test="cell0-5"
  >
    <span
      class=" "
    >
      5 TB
    </span>
  </div>
  <div
    class="c3 c4"
    data-condition="used"
    data-test="cell0-6"
  >
    <span
      class=" "
    >
      used
    </span>
  </div>
  <div
    class="c5"
    data-condition="used"
    data-test="cell0-7"
  >
    <a
      class="c9"
      href="https://url1"
      target="_blank"
    >
      WD 5 TB My Passport
    </a>
  </div>
  <div
    class="c6 c7"
    data-condition="used"
    data-test="cell0-8"
  >
    <span
      class=" c10"
    >
      <span
        class=""
      >
        over 21 years
      </span>
    </span>
  </div>
  <div
    class="c3 c4"
    data-condition="used"
    data-test="cell1-1"
  >
    <span
      class=" "
    >
      -
    </span>
  </div>
  <div
    class="c3 c4"
    data-condition="used"
    data-test="cell1-2"
  >
    <img
      class="c8"
      height="60"
      src="https://image2"
      width="60"
    />
  </div>
  <div
    class="c3 c4"
    data-condition="used"
    data-test="cell1-3"
  >
    <span
      class=" "
    >
      £ 22.62
    </span>
  </div>
  <div
    class="c3 c4"
    data-condition="used"
    data-test="cell1-4"
  >
    <span
      class=" "
    >
      £ 180.9911
    </span>
  </div>
  <div
    class="c3 c4"
    data-condition="used"
    data-test="cell1-5"
  >
    <span
      class=" "
    >
      8 TB
    </span>
  </div>
  <div
    class="c3 c4"
    data-condition="used"
    data-test="cell1-6"
  >
    <span
      class=" "
    >
      used
    </span>
  </div>
  <div
    class="c5"
    data-condition="used"
    data-test="cell1-7"
  >
    <a
      class="c9"
      href="https://url2"
      target="_blank"
    >
      Seagate 8 TB
    </a>
  </div>
  <div
    class="c6 c7"
    data-condition="used"
    data-test="cell1-8"
  >
    <span
      class=" c10"
    >
      <span
        class=""
      >
        over 21 years
      </span>
    </span>
  </div>
  <div
    class="c3 c4"
    data-condition="used"
    data-test="cell2-1"
  >
    <span
      class=" "
    >
      -
    </span>
  </div>
  <div
    class="c3 c4"
    data-condition="used"
    data-test="cell2-2"
  >
    <span
      class=" c10"
    >
      [no image]
    </span>
  </div>
  <div
    class="c3 c4"
    data-condition="used"
    data-test="cell2-3"
  >
    <span
      class=" "
    >
      £ 22.62
    </span>
  </div>
  <div
    class="c3 c4"
    data-condition="used"
    data-test="cell2-4"
  >
    <span
      class=" "
    >
      £ 180.99
    </span>
  </div>
  <div
    class="c3 c4"
    data-condition="used"
    data-test="cell2-5"
  >
    <span
      class=" "
    >
      8 TB
    </span>
  </div>
  <div
    class="c3 c4"
    data-condition="used"
    data-test="cell2-6"
  >
    <span
      class=" "
    >
      used
    </span>
  </div>
  <div
    class="c5"
    data-condition="used"
    data-test="cell2-7"
  >
    <a
      class="c9"
      href="https://url2"
      target="_blank"
    >
      Seagate 8 TB (no image, no date)
    </a>
  </div>
  <div
    class="c6 c7"
    data-condition="used"
    data-test="cell2-8"
  >
    <span
      class=" c10"
    >
      <span
        class=""
      >
        about 30 years
      </span>
    </span>
  </div>
  <div
    class="c3 c4"
    data-condition="new"
    data-test="cell3-1"
  >
    <span
      class=" "
    >
      -
    </span>
  </div>
  <div
    class="c3 c4"
    data-condition="new"
    data-test="cell3-2"
  >
    <img
      class="c8"
      height="60"
      src="https://image1"
      width="60"
    />
  </div>
  <div
    class="c3 c4"
    data-condition="new"
    data-test="cell3-3"
  >
    <span
      class=" "
    >
      £ 24.12
    </span>
  </div>
  <div
    class="c3 c4"
    data-condition="new"
    data-test="cell3-4"
  >
    <span
      class=" "
    >
      £ 120.59
    </span>
  </div>
  <div
    class="c3 c4"
    data-condition="new"
    data-test="cell3-5"
  >
    <span
      class=" "
    >
      5 TB
    </span>
  </div>
  <div
    class="c3 c4"
    data-condition="new"
    data-test="cell3-6"
  >
    <span
      class=" "
    >
      new
    </span>
  </div>
  <div
    class="c5"
    data-condition="new"
    data-test="cell3-7"
  >
    <a
      class="c9"
      href="https://url1"
      target="_blank"
    >
      WD 5 TB My Passport
    </a>
  </div>
  <div
    class="c6 c7"
    data-condition="new"
    data-test="cell3-8"
  >
    <span
      class=" c10"
    >
      <span
        class=""
      >
        over 21 years
      </span>
    </span>
  </div>
</div>
`;
